### PR TITLE
:sparkles: PlacementDecision Logic

### PR DESCRIPTION
**Pull request by vMaroon**
Recreation of https://github.com/kubestellar/kubestellar/pull/1633

After going through internal reviews and merging to main, another PR on removing marked redundant code (from previous implementation) will be opened.

### CODE REVIEW

File: pkg/placement/controller.go [hunk 1]
```diff
@@ -75,17 +75,20 @@ var excludedResourceNames = map[string]bool{
 type Controller struct {
 	ctx              context.Context
 	logger           logr.Logger
-	ocmClient        client.Client
+	ocmClient        client.Client // TODO (maroon): this should be deleted when transport is ready
```
Comment (nirrozenbaum) [inline pkg/placement/controller.go:78]: transport is ready.. when do you think we should remove ocm client?
do you want to start testing using the updated transport?

File: pkg/placement/controller.go [hunk 2]
```diff
@@ -75,17 +75,20 @@ var excludedResourceNames = map[string]bool{
 type Controller struct {
```
Comment (nirrozenbaum) [inline pkg/placement/controller.go:75]: general comment, not on your code but on this file (probably Paolo's code):
I didn't understand why the excluded versions and resources are stored in map and not in a set (or array, or list).
the question if resource/version should be excluded is if set contains it or not. there is no meaning for the true/false value here.

File: pkg/placement/controller.go [hunk 3]
```diff
@@ -75,17 +75,20 @@ var excludedResourceNames = map[string]bool{
...
 	ctx              context.Context
 	logger           logr.Logger
-	ocmClient        client.Client
+	ocmClient        client.Client // TODO (maroon): this should be deleted when transport is ready
 	dynamicClient    *dynamic.DynamicClient
 	kubernetesClient *kubernetes.Clientset
```
Comment (nirrozenbaum) [inline pkg/placement/controller.go:80]: use `kubernetes.Interface` to allow unit tests with fake

File: pkg/placement/controller.go [hunk 4]
```diff
@@ -75,17 +75,20 @@ var excludedResourceNames = map[string]bool{
 type Controller struct {
 	ctx              context.Context
 	logger           logr.Logger
-	ocmClient        client.Client
+	ocmClient        client.Client // TODO (maroon): this should be deleted when transport is ready
 	dynamicClient    *dynamic.DynamicClient
```
Comment (nirrozenbaum) [inline pkg/placement/controller.go:79]: use `dynamic.Interface` to allow unit tests with fake

File: pkg/placement/controller.go [hunk 5]
```diff
@@ -75,17 +75,20 @@ var excludedResourceNames = map[string]bool{
...
 	kubernetesClient *kubernetes.Clientset
 	extClient        *apiextensionsclientset.Clientset
```
Comment (nirrozenbaum) [inline pkg/placement/controller.go:81]: didn't check but I assume there is Interface as well..
that was actually important comment I got from Mike in my PR.
after code is ready and manually tested we will probably want to add unit tests

File: pkg/placement/controller.go [hunk 6]
```diff
@@ -112,18 +115,21 @@ func NewController(mgr ctrlm.Manager, wdsRestConfig *rest.Config, imbsRestConfig
...
+		logger:                    mgr.GetLogger(),
+		ocmClient:                 ocmClient,
+		dynamicClient:             dynamicClient,
+		kubernetesClient:          kubernetesClient,
+		extClient:                 extClient,
+		listers:                   make(map[string]*cache.GenericLister),
+		informers:                 make(map[string]*cache.SharedIndexInformer),
+		stoppers:                  make(map[string]chan struct{}),
+		placementDecisionResolver: NewPlacementDecisionResolver(gvkGvrMapper),
+		gvkGvrMapper:              gvkGvrMapper,
```
Comment (nirrozenbaum) [inline pkg/placement/controller.go:131]: remove variable.. `gvkGvrMapper: util.NewGvkGvrMapper()`

File: pkg/placement/controller.go [hunk 7]
```diff
@@ -112,18 +115,21 @@ func NewController(mgr ctrlm.Manager, wdsRestConfig *rest.Config, imbsRestConfig
```
Comment (nirrozenbaum) [inline pkg/placement/controller.go:115]: another comment that is probably leftover from Paolo's code.
receiving in the NewController args from type `rest.config` is not testable. 
clients should be created in the cmd and received as args in pkg.
this way NewController should get arguments from types: `dynamic.Interface`,`kubernetes.Interface`,`apiextensionclientset.Interface` which will allow to test this code with fake implementations

File: pkg/placement/controller.go [hunk 8]
```diff
@@ -75,17 +75,20 @@ var excludedResourceNames = map[string]bool{
 type Controller struct {
 	ctx              context.Context
```
Comment (nirrozenbaum) [inline pkg/placement/controller.go:76]: context should not be part of the controller struct. it should be received in Start function as arg.

File: pkg/placement/controller.go [hunk 9]
```diff
@@ -75,17 +75,20 @@ var excludedResourceNames = map[string]bool{
...
 	logger           logr.Logger
-	ocmClient        client.Client
+	ocmClient        client.Client // TODO (maroon): this should be deleted when transport is ready
 	dynamicClient    *dynamic.DynamicClient
 	kubernetesClient *kubernetes.Clientset
 	extClient        *apiextensionsclientset.Clientset
 	listers          map[string]*cache.GenericLister
-	gvksMap          map[string]*schema.GroupVersionResource
+	gvkGvrMapper     util.GvkGvrMapper
 	informers        map[string]*cache.SharedIndexInformer
```
Comment (nirrozenbaum) [inline pkg/placement/controller.go:84]: why do you have pointer to `cache.SharedIndexInformer` in the map and not just `cache.SharedIndexInformer`?
This is an interface (it's already kind of a pointer).. putting pointer to interface looks a bit strange.

File: pkg/placement/controller.go [hunk 10]
```diff
@@ -75,17 +75,20 @@ var excludedResourceNames = map[string]bool{
 type Controller struct {
 	ctx              context.Context
 	logger           logr.Logger
-	ocmClient        client.Client
+	ocmClient        client.Client // TODO (maroon): this should be deleted when transport is ready
 	dynamicClient    *dynamic.DynamicClient
 	kubernetesClient *kubernetes.Clientset
 	extClient        *apiextensionsclientset.Clientset
 	listers          map[string]*cache.GenericLister
```
Comment (nirrozenbaum) [inline pkg/placement/controller.go:82]: same comment as in `SharedIndexInformer` map

File: pkg/placement/controller.go (hunk 1 above)
Comment (vMaroon) [reply pkg/placement/controller.go:78]: once this PR is in, I'll open another that removes the marked code and test e2e.

File: pkg/placement/controller.go (hunk 6 above)
Comment (vMaroon) [reply pkg/placement/controller.go:131]: it needs to be the same mapper for both

Comment (nirrozenbaum) [reply pkg/placement/controller.go:131]: missed the second. you're right.

**Comment by vMaroon** [comment]
merging to open against KS repo

**Pull request merged**